rslib/src/typeanswer.rs

```diff
@@ -5,113 +5,160 @@ use std::borrow::Cow;
 use unic_ucd_category::GeneralCategory;
 
 use crate::card_rendering::strip_av_tags;
 use crate::text::normalize_to_nfc;
 use crate::text::strip_html;
 
-lazy_static! {
-    static ref LINEBREAKS: Regex = Regex::new(
+static LINEBREAKS: Lazy<Regex> = Lazy::new(|| {
+    Regex::new(
         r"(?six)
         (
             \n
             |
             <br\s?/?>
             |
             </?div>
-        )+
-    "
+        )+",
     )
-    .unwrap();
+    .unwrap()
+});
+
+macro_rules! format_typeans {
+    ($typeans:expr) => {
+        format!("<code id=typeans>{}</code>", $typeans)
+    };
 }
 
-struct DiffContext {
-    expected: Vec<char>,
+// Public API
+pub fn compare_answer(expected: &str, provided: &str) -> String {
+    if provided.is_empty() {
+        format_typeans!(htmlescape::encode_minimal(expected))
+    } else {
+        Diff::new(expected, provided).to_html()
+    }
+}
+
+struct Diff {
     provided: Vec<char>,
+    expected: Vec<char>,
+    expected_original: String,
 }
 
-impl DiffContext {
+impl Diff {
     fn new(expected: &str, provided: &str) -> Self {
-        DiffContext {
-            provided: prepare_provided(provided).chars().collect_vec(),
-            expected: prepare_expected(expected).chars().collect_vec(),
+        Self {
+            provided: normalize_to_nfc(provided).chars().collect(),
+            expected: normalize_to_nfc(&Self::prepare_expected(expected))
+                .chars()
+                .collect(),
+            expected_original: expected.to_string(),
         }
     }
 
-    fn slice_expected(&self, opcode: &Opcode) -> String {
-        self.expected[opcode.second_start..opcode.second_end]
-            .iter()
-            .cloned()
-            .collect()
-    }
-
-    fn slice_provided(&self, opcode: &Opcode) -> String {
-        self.provided[opcode.first_start..opcode.first_end]
-            .iter()
-            .cloned()
-            .collect()
+    // Main Functions
+    fn to_html(&self) -> String {
+        if self.provided == self.expected {
+            format_typeans!(format!(
+                "<span class=typeGood>{}</span>",
+                self.expected_original
+            ))
+        } else {
+            let output = self.to_tokens();
+            let provided_html = Self::render_tokens(&output.provided_tokens);
+            let expected_html = Self::render_tokens(&output.expected_tokens);
+
+            format_typeans!(format!(
+                "{provided_html}<br><span id=typearrow>&darr;</span><br>{expected_html}"
+            ))
+        }
     }
 
-    fn to_tokens(&self) -> DiffOutput {
+    fn to_tokens(&self) -> DiffTokens {
         let mut matcher = SequenceMatcher::new(&self.provided, &self.expected);
-        let opcodes = matcher.get_opcodes();
-        let mut provided = vec![];
-        let mut expected = vec![];
-        for opcode in opcodes {
+        let mut provided_tokens = Vec::new();
+        let mut expected_tokens = Vec::new();
```

Comment:
The class already includes Tokens in the name, so using _tokens for the property names feels a bit redundant.

Reply:
`provided` and `expected` were used repeatedly as identifier for different variables in that same file. So I figured the addition worth it for clarity.